tiledb/ml/readers/pytorch.py

```diff
@@ -86,15 +90,16 @@ def __iter__(self) -> Iterator[DataType]:
             iter_end = min(iter_start + per_worker, rows)
 
         # Loop over batches
-        for offset in range(iter_start, iter_end, self.batch_size):
-            x_batch, y_batch = run_io_tasks_in_parallel(
-                (self.x, self.y), self.batch_size, offset
-            )
-
-            # Yield the next training batch
-            yield tuple(x_batch[attr] for attr in self.x_attribute_names) + tuple(
-                y_batch[attr] for attr in self.y_attribute_names
-            )
+        with ThreadPoolExecutor(max_workers=2) as self.executor:
+            for offset in range(iter_start, iter_end, self.batch_size):
+                x_batch, y_batch = self.run_io_tasks_in_parallel(
```

Comment:
When I suggested a mixin I meant create `self.executor` once in `__init__` (either in this class or preferably in `ParallelIOMixin` itself). Creating it in `__iter__` is fine (assuming it's not called too often) and has the benefit of using a context manager for cleanup, but in this case you don't need an instance attribute and a mixin; just change `run_io_tasks_in_parallel` to a function that takes `executor` as parameter.

Reply:
I would also create it once in `ParallelIOMixin` `__init__`, but I guess @ktsitsi could have pickling issues coming from PyTorch. Is this the case @ktsitsi ?

Reply:
Yeap @gsakkis I did exactly that at first place...but I faced an error due to `pytorch`.  I am posting the error below just for reference. Then I also considered what you also pointed out on the benefits of a `context manager` for the cleanup and that is why I eventually implemented it that way. If you have any idea about the error below I am totally open to discuss.

```
obj = <Process name='Process-1' parent=50591 initial daemon>
file = <_io.BytesIO object at 0x169f0f180>, protocol = None

    def dump(obj, file, protocol=None):
        '''Replacement for pickle.dump() using ForkingPickler.'''
>       ForkingPickler(file, protocol).dump(obj)
E       TypeError: cannot pickle '_queue.SimpleQueue' object

```

Reply:
I prefer it in `__iter__`, just replace the `self.executor` attribute with a local variable and change the mixin to a plain function that takes an executor (or even inline the whole thing, it's just 2 lines).